File: lib/hdbfs/hooks.py
import sys

from defs import *

_PRE_COMMIT_HOOKS = []
_POST_COMMIT_HOOKS = []

def add_pre_commit_hook( h ):
    global _PRE_COMMIT_HOOKS
    _PRE_COMMIT_HOOKS.append( h )

def add_post_commit_hook( h ):
    global _POST_COMMIT_HOOKS
    _POST_COMMIT_HOOKS.append( h )

def trigger_pre_commit_hooks( db, is_rollback ):
    global _PRE_COMMIT_HOOKS

    for h in _PRE_COMMIT_HOOKS:
        try:
            h( db, is_rollback )
        except:
            LOG.warning( 'Pre commit hook "%s" failed: %s',
                         str( h ), str( sys.exc_info()[1] ) )

def trigger_post_commit_hooks( db, is_rollback ):
    global _POST_COMMIT_HOOKS

    for h in _POST_COMMIT_HOOKS:
        try:
            h( db, is_rollback )
        except:
            LOG.warning( 'Post commit hook "%s" failed: %s',
                         str( h ), str( sys.exc_info()[1] ) )

File: lib/src/main/kotlin/ca/_4haven/higu/hdbfs/Hooks.kt
package ca._4haven.higu.hdbfs

typealias Hook = ( db: Database, is_rollback: Boolean ) -> Unit

class Hooks( val db: Database ) {

    private var _PRE_COMMIT_HOOKS = mutableListOf<Hook>()
    private var _POST_COMMIT_HOOKS = mutableListOf<Hook>()

    fun add_pre_commit_hook( h: Hook ) {
        this._PRE_COMMIT_HOOKS.add( h )
    }

    fun add_post_commit_hook( h: Hook ) {
        _POST_COMMIT_HOOKS.add( h )
    }

    fun trigger_pre_commit_hooks( is_rollback: Boolean ) {
        _PRE_COMMIT_HOOKS.forEach { h ->
            try {
                h( db, is_rollback )
            } catch( ex: Exception ) {
                Log.warning( "Pre commit hook \"${h}\" failed: ${ex}" )
            }
        }
    }

    fun trigger_post_commit_hooks( is_rollback: Boolean ) {
        _POST_COMMIT_HOOKS.forEach { h ->
            try {
                h( db, is_rollback )
            } catch( ex: Exception ) {
                Log.warning( "Post commit hook \"${h}\" failed: ${ex}" )
            }
        }
    }
}